README.md
[![Pub Version](https://img.shields.io/pub/v/dialog_context?color=%2302569B&label=pub&logo=flutter)](https://pub.dev/packages/dialog_context) ![PRs Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)

### Show Dialog, BottomSheet, ModalBottomSheet and SnackBar with no need BuildContext.
To be discontinued, use the OneContext package: [one_context](https://pub.dev/packages/one_context)


## https://pub.dev/packages/one_context

```dart
// example snackBar
DialogContext().showSnackBar(
    snackBar: SnackBar(content: Text('My awesome snackBar!'))
);
```

```dart
// example dialog
DialogContext().showDialog(
    builder: (context) => AlertDialog( 
        title: new Text("The Title"),
        content: new Text("The Body"),
    )
);
```

```dart
// example bottomSheet
DialogContext().showBottomSheet(
    builder: (context) => Container(
    alignment: Alignment.topCenter,
    height: 200,
    child: IconButton(
        icon: Icon(Icons.arrow_drop_down),
        iconSize: 50,
        color: Colors.white,
        onPressed: () => Navigator.pop(context)),
    ),
);
```

## Important: Use the DialogContext().builder in MaterialApp
```dart
MaterialApp(
  builder: DialogContext().builder, /// important: Use [DialogContext().builder] builder here!
  home: MyHomePage(title: 'Dialogs without BuildContext'),
);
```

#### Add dependency
```yaml
  dependencies:
    dialog_context: ^0.1.1
```


CHANGELOG.md
## [0.1.1] - Refactors

## [0.1.0] - Initial release
* add showDialog
* add showModalBottomSheet
* add showBottomSheet
* add showSnackBar
* add dismissSnackBar


lib/src/controllers/dialog_context.dart
import 'package:dialog_context/src/components/dialog_context_widget.dart';
import 'package:flutter/material.dart';

typedef Future<T> DialogFuture<T>(
    {bool barrierDismissible,
    Widget Function(BuildContext) builder,
    bool useRootNavigator});
typedef Future<T> DialogModalSheet<T>(
    {Widget Function(BuildContext) builder,
    Color backgroundColor,
    double elevation,
    ShapeBorder shape,
    Clip clipBehavior,
    bool isScrollControlled,
    bool useRootNavigator,
    bool isDismissible});
typedef Widget DialogBuilder(BuildContext context);
typedef ScaffoldFeatureController<SnackBar,
    SnackBarClosedReason> DialogSnackBar(SnackBar snackBar);
typedef PersistentBottomSheetController<T> DialogBottomSheet<T>(
    {Widget Function(BuildContext) builder,
    Color backgroundColor,
    double elevation,
    ShapeBorder shape,
    Clip clipBehavior});

class DialogContext {
  BuildContext context;

  DialogFuture _showDialog;
  DialogModalSheet _showModalBottomSheet;
  DialogSnackBar _showSnackBar;
  DialogBottomSheet _showBottomSheet;

  DialogContext._private();
  static DialogContext _instance = DialogContext._private();
  factory DialogContext() => _instance;

  Widget builder(BuildContext context, Widget widget) {
    return Navigator(
      onGenerateRoute: (settings) => MaterialPageRoute(
        builder: (context) => DialogContextWidget(
          child: widget,
        ),
      ),
    );
  }

  void registerDialogCallback(
      {DialogFuture showDialog,
      DialogModalSheet showModalBottomSheet,
      DialogSnackBar showSnackBar,
      DialogBottomSheet showBottomSheet}) {
    _showDialog = showDialog;
    _showSnackBar = showSnackBar;
    _showModalBottomSheet = showModalBottomSheet;
    _showBottomSheet = showBottomSheet;
  }

  Future<T> showDialog<T>(
      {@required Widget Function(BuildContext) builder,
      bool barrierDismissible = true,
      bool useRootNavigator = true}) async {
    assert(builder != null);
    return await _showDialog(
      builder: builder,
      barrierDismissible: barrierDismissible,
      useRootNavigator: useRootNavigator,
    );
  }

  void dismissSnackBar() => Scaffold.of(context).hideCurrentSnackBar();
  ScaffoldFeatureController<SnackBar, SnackBarClosedReason> showSnackBar(
      {@required SnackBar snackBar}) {
    assert(snackBar != null);
    return _showSnackBar(snackBar);
  }

  Future<T> showModalBottomSheet<T>(
      {@required Widget Function(BuildContext) builder,
      Color backgroundColor,
      double elevation,
      ShapeBorder shape,
      Clip clipBehavior,
      bool isScrollControlled = false,
      bool useRootNavigator = false,
      bool isDismissible = true}) async {
    assert(builder != null);
    return await _showModalBottomSheet(
        builder: builder,
        backgroundColor: backgroundColor,
        clipBehavior: clipBehavior,
        elevation: elevation,
        isDismissible: isDismissible,
        isScrollControlled: isScrollControlled,
        shape: shape,
        useRootNavigator: useRootNavigator);
  }

  PersistentBottomSheetController<T> showBottomSheet<T>(
      {@required Widget Function(BuildContext) builder,
      Color backgroundColor,
      double elevation,
      ShapeBorder shape,
      Clip clipBehavior}) {
    assert(builder != null);
    return _showBottomSheet(
        builder: builder,
        backgroundColor: backgroundColor,
        elevation: elevation,
        shape: shape,
        clipBehavior: clipBehavior);
  }
}


lib/src/components/dialog_context_widget.dart
import 'package:dialog_context/src/controllers/dialog_context.dart';
import 'package:flutter/material.dart';

class DialogContextWidget extends StatefulWidget {
  final Widget child;
  DialogContextWidget({Key key, this.child}) : super(key: key);
  _DialogContextWidgetState createState() => _DialogContextWidgetState();
}

class _DialogContextWidgetState extends State<DialogContextWidget> {
  @override
  void initState() {
    super.initState();
    DialogContext().registerDialogCallback(
        showDialog: _showDialog,
        showSnackBar: _showSnackBar,
        showModalBottomSheet: _showModalBottomSheet,
        showBottomSheet: _showBottomSheet);
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Builder(
        builder: (scaffoldContext) {
          DialogContext().context = scaffoldContext;
          return widget.child;
        },
      ),
    );
  }

  Future<T> _showDialog<T>(
          {bool barrierDismissible = true,
          Widget Function(BuildContext) builder,
          bool useRootNavigator = true}) =>
      showDialog<T>(
        context: context,
        builder: (context) => builder(context),
        barrierDismissible: barrierDismissible,
        useRootNavigator: useRootNavigator,
      );

  ScaffoldFeatureController<SnackBar, SnackBarClosedReason> _showSnackBar(
          SnackBar snackBar) =>
      Scaffold.of(DialogContext().context).showSnackBar(snackBar);

  Future<T> _showModalBottomSheet<T>(
      {Widget Function(BuildContext) builder,
      Color backgroundColor,
      double elevation,
      ShapeBorder shape,
      Clip clipBehavior,
      bool isScrollControlled = false,
      bool useRootNavigator = false,
      bool isDismissible = true}) {
    return showModalBottomSheet<T>(
      context: context,
      builder: builder,
      backgroundColor: backgroundColor,
      clipBehavior: clipBehavior,
      elevation: elevation,
      isDismissible: isDismissible,
      isScrollControlled: isScrollControlled,
      shape: shape,
      useRootNavigator: useRootNavigator,
    );
  }

  PersistentBottomSheetController<T> _showBottomSheet<T>(
      {@required Widget Function(BuildContext) builder,
      Color backgroundColor,
      double elevation,
      ShapeBorder shape,
      Clip clipBehavior}) {
    return showBottomSheet<T>(
        context: DialogContext().context,
        builder: builder,
        backgroundColor: backgroundColor,
        elevation: elevation,
        shape: shape,
        clipBehavior: clipBehavior);
  }
}


example/lib/main.dart
import 'package:dialog_context/dialog_context.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      builder: DialogContext().builder,

      /// important: Use [DialogContext().builder] builder
      title: 'Dialogs without BuildContext',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Dialogs without BuildContext'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(
        alignment: Alignment.center,
        padding: EdgeInsets.all(16),
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.stretch,
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            RaisedButton(
              child: Text('Show SnackBar'),
              onPressed: () {
                DialogContext().dismissSnackBar(); // Dismiss snackbar
                DialogContext().showSnackBar(
                    snackBar: SnackBar(
                  content: Text(
                    'My awesome snackBar!',
                    textAlign: TextAlign.center,
                    style: Theme.of(context)
                        .textTheme
                        .title
                        .copyWith(color: Colors.white),
                  ),
                  action: SnackBarAction(label: 'DISMISS', onPressed: () {}),
                ));
              },
            ),
            RaisedButton(
              child: Text('Show Dialog'),
              onPressed: () async {
                var result = await DialogContext().showDialog<String>(
                    // Don't need context to show alertDialog
                    builder: (context) => AlertDialog(
                          title: new Text("The Title"),
                          content: new Text("The Body"),
                          actions: <Widget>[
                            new FlatButton(
                                child: new Text("OK"),
                                onPressed: () =>
                                    Navigator.of(context).pop('ok')),
                            new FlatButton(
                                child: new Text("CANCEL"),
                                onPressed: () =>
                                    Navigator.of(context).pop('cancel')),
                          ],
                        ));
                print(result);
              },
            ),
            RaisedButton(
              child: Text('Show modalBottomSheet'),
              onPressed: () async {
                var result = await DialogContext().showModalBottomSheet<String>(
                    builder: (context) => Container(
                          child: Column(
                            mainAxisSize: MainAxisSize.min,
                            children: <Widget>[
                              ListTile(
                                  leading: Icon(Icons.music_note),
                                  title: Text('Music'),
                                  onTap: () =>
                                      Navigator.of(context).pop('Music')),
                              ListTile(
                                leading: Icon(Icons.videocam),
                                title: Text('Video'),
                                onTap: () => Navigator.of(context).pop('Video'),
                              ),
                              SizedBox(height: 45)
                            ],
                          ),
                        ));
                print(result);
              },
            ),
            RaisedButton(
              child: Text('Show bottomSheet'),
              onPressed: () {
                DialogContext().showBottomSheet(
                  builder: (context) => Container(
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.all(Radius.circular(10)),
                      color: Colors.red,
                    ),
                    margin: EdgeInsets.all(16),
                    alignment: Alignment.topCenter,
                    height: 200,
                    child: IconButton(
                        icon: Icon(Icons.arrow_drop_down),
                        iconSize: 50,
                        color: Colors.white,
                        onPressed: () => Navigator.pop(context)),
                  ),
                );
              },
            ),
            RaisedButton(
                child: Text('Show and await input'),
                onPressed: () async {
                  switch (await DialogContext().showDialog<int>(
                      barrierDismissible: false,
                      builder: (BuildContext context) {
                        return SimpleDialog(
                          shape: RoundedRectangleBorder(
                              borderRadius:
                                  BorderRadius.all(Radius.circular(10))),
                          title: const Text('Select assignment'),
                          children: <Widget>[
                            SimpleDialogOption(
                              onPressed: () {
                                Navigator.pop(context, 1);
                              },
                              child: const Text('Number 1'),
                            ),
                            SimpleDialogOption(
                              onPressed: () {
                                Navigator.pop(context, 2);
                              },
                              child: const Text('Number 2'),
                            ),
                          ],
                        );
                      })) {
                    case 1:
                      print('number one');
                      break;
                    case 2:
                      print('number two');
                      break;
                  }
                }),
          ],
        ),
      ),
    );
  }
}
